src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.BatchingDocumentCollection.cs

```diff
@@ -537,40 +543,63 @@ public void ReorderFiles(ImmutableArray<string> filePaths)
-                documentIds: _documentsAddedInBatch.Select(d => d.Id));
-
-            foreach (var documentInfo in _documentsAddedInBatch)
+            UpdateSolutionForBatch(solutionChanges, documentFileNamesAdded, documentsToOpen, addDocuments,
+                addDocumentChangeKind, removeDocuments, removeDocumentChangeKind, _project.Id, _documentsAddedInBatch.ToImmutableArray(),
+                _documentsRemovedInBatch.ToImmutableArray(), _orderedDocumentsInBatch,
+                (documentId) => _sourceTextContainersToDocumentIds.GetKeyOrDefault(documentId));
+
+            static void UpdateSolutionForBatch(
+                SolutionChangeAccumulator solutionChanges,
+                ImmutableArray<string>.Builder documentFileNamesAdded,
+                List<(DocumentId documentId, SourceTextContainer textContainer)> documentsToOpen,
+                Func<Solution, ImmutableArray<DocumentInfo>, Solution> addDocuments,
+                WorkspaceChangeKind addDocumentChangeKind,
+                Func<Solution, ImmutableArray<DocumentId>, Solution> removeDocuments,
+                WorkspaceChangeKind removeDocumentChangeKind,
+                ProjectId projectId,
+                ImmutableArray<DocumentInfo> documentsAddedInBatch,
+                ImmutableArray<DocumentId> documentsRemovedInBatch,
+                ImmutableList<DocumentId>? orderedDocumentsInBatch,
+                Func<DocumentId, SourceTextContainer?> getContainer)
             {
-                Contract.ThrowIfNull(documentInfo.FilePath, "We shouldn't be adding documents without file paths.");
-                documentFileNamesAdded.Add(documentInfo.FilePath);
+                // Document adding...
+                solutionChanges.UpdateSolutionForDocumentAction(
+                    newSolution: addDocuments(solutionChanges.Solution, documentsAddedInBatch),
+                    changeKind: addDocumentChangeKind,
+                    documentIds: documentsAddedInBatch.Select(d => d.Id));
 
-                if (_sourceTextContainersToDocumentIds.TryGetKey(documentInfo.Id, out var textContainer))
+                foreach (var documentInfo in documentsAddedInBatch)
                 {
-                    documentsToOpen.Add((documentInfo.Id, textContainer));
+                    Contract.ThrowIfNull(documentInfo.FilePath, "We shouldn't be adding documents without file paths.");
+                    documentFileNamesAdded.Add(documentInfo.FilePath);
+
+                    var textContainer = getContainer(documentInfo.Id);
+                    if (textContainer != null)
+                    {
+                        documentsToOpen.Add((documentInfo.Id, textContainer));
+                    }
                 }
-            }
 
-            ClearAndZeroCapacity(_documentsAddedInBatch);
+                // Document removing...
+                solutionChanges.UpdateSolutionForRemovedDocumentAction(removeDocuments(solutionChanges.Solution, [.. documentsRemovedInBatch]),
```

Comment:
you shouldn't need the `[.. documentsRemovedInBatch]` anymore, since you already  have an immutable copy, so it should be safe to pass withou tanother copy.